### PR TITLE
Separate remotely downloaded snapshot archives

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

This is yet another approach, from https://github.com/solana-labs/solana/pull/23455, to address the https://github.com/solana-labs/solana/issues/23452

#### Summary of Changes
1. Remotely downloaded snapshot archives are stored in `remote` subdirectory.
2. skip cleaning and shrinking for local snapshot archives.

Fixes #

### CODE REVIEW

File: runtime/src/snapshot_archive_info.rs [hunk 1]
```diff
@@ -65,6 +65,10 @@ impl FullSnapshotArchiveInfo {
     pub(crate) fn new(snapshot_archive_info: SnapshotArchiveInfo) -> Self {
         Self(snapshot_archive_info)
     }
+
+    pub fn is_remote(&self) -> bool {
+        self.path().parent().unwrap().ends_with("remote")
```
Comment (sakridge) [inline runtime/src/snapshot_archive_info.rs:70]: Would be nice to consolidate the “remote” constant.

File: download-utils/src/lib.rs [hunk 2]
```diff
@@ -274,16 +274,19 @@ pub fn download_snapshot_archive<'a, 'b>(
         ArchiveFormat::TarBzip2,
         ArchiveFormat::Tar, // `solana-test-validator` creates uncompressed snapshots
     ] {
+        let snapshot_archives_remote_dir = snapshot_archives_dir.to_path_buf().join("remote");
+        fs::create_dir_all(snapshot_archives_remote_dir.to_path_buf()).unwrap();
```
Comment (brooksprumo) [inline download-utils/src/lib.rs]: Will this `.unwrap()` assert if the directory already exists? If the archive format is not a TarZstd, then the directory would've already been created, right? Maybe this code needs to be hoisted out of the loop.

Or alternatively the `snapshot_archives_dir` parameter could already be the remote directory?

File: runtime/src/snapshot_utils.rs [hunk 3]
```diff
@@ -1139,54 +1139,63 @@ pub(crate) fn parse_incremental_snapshot_archive_filename(
...
-                err,
-                snapshot_archives_dir.as_ref().display()
+                e,
+                dir.display()
             );
-            vec![]
+            return;
+        }
+
+        for entry in entry_iter.unwrap() {
```
Comment (sakridge) [inline runtime/src/snapshot_utils.rs:1155]: It will look now in every folder and sub-folder of the ledger directory now? Is that right? Does it walk accounts folder? Because we have a lot of files there.

File: runtime/src/snapshot_archive_info.rs [hunk 4]
```diff
@@ -65,6 +65,10 @@ impl FullSnapshotArchiveInfo {
...
+    }
```
Comment (brooksprumo) [inline runtime/src/snapshot_archive_info.rs]: 1. I'd rather see this function return `false` in case `.parent()` fails, instead of crashing.

2. I think this fn should be part of the SnapshotArchiveInfoGetter trait, instead of just on the `FullSnapshotArchiveInfo` struct.

File: runtime/src/snapshot_utils.rs (hunk 3 above)
Comment (HaoranYi) [reply runtime/src/snapshot_utils.rs:1156]: good catch. yes, it does. i will limit it to only on `remote` sub directory.

File: runtime/src/snapshot_utils.rs [hunk 5]
```diff
@@ -1139,54 +1140,63 @@ pub(crate) fn parse_incremental_snapshot_archive_filename(
...
 
-/// Get a list of the full snapshot archives in a directory
-pub fn get_full_snapshot_archives<P>(snapshot_archives_dir: P) -> Vec<FullSnapshotArchiveInfo>
-where
-    P: AsRef<Path>,
-{
-    match fs::read_dir(&snapshot_archives_dir) {
-        Err(err) => {
+/// Walk down the snapshot archive directory recursively to collect snapshot archive file info
+fn get_snapshot_archives<T>(dir: &Path, cb: fn(PathBuf) -> Result<T>, output: &mut Vec<T>) {
```
Comment (brooksprumo) [inline runtime/src/snapshot_utils.rs:1144]: Could this fn be simplified? Getting all the snapshot archives just needs to check an additional "remote" directory in addition to the normal one. And we could go back to returning the results instead of requiring the `&mut` parameter.

File: download-utils/src/lib.rs (hunk 2 above)
Comment (HaoranYi) [reply download-utils/src/lib.rs]: https://doc.rust-lang.org/std/fs/fn.create_dir_all.html
Based on the documentation, it shouldn't assert. unlike `fs::create_dir`.
Yes, we should move the code out of the loop.

File: runtime/src/snapshot_utils.rs (hunk 5 above)
Comment (HaoranYi) [reply runtime/src/snapshot_utils.rs:1144]: yeah. only need to look into remote sub directory. Commented by @sakridge as well.

File: runtime/src/snapshot_archive_info.rs (hunk 4 above)
Comment (HaoranYi) [reply runtime/src/snapshot_archive_info.rs]: yeah.

**Comment by HaoranYi** [comment]
The following local cluster test failed. 
test test_incremental_snapshot_download_with_crossing_full_snapshot_interval_at_startup 
I will look into it today. 
@brooksprumo @sakridge Any suggestion and hint to debug the failure?

**Comment by sakridge** [comment]
> 

Not sure off the top of my head. What's the assert or failure and does the test print anything with `RUST_LOG=info` enabled?

File: runtime/src/blockhash_queue.rs [hunk 6]
```diff
@@ -130,7 +130,7 @@ impl BlockhashQueue {
     )]
     #[allow(deprecated)]
     pub fn get_recent_blockhashes(&self) -> impl Iterator<Item = recent_blockhashes::IterItem> {
-        (&self.ages).iter().map(|(k, v)| {
+        (self.ages).iter().map(|(k, v)| {
```
Comment (brooksprumo) [inline runtime/src/blockhash_queue.rs:133]: nit: This change doesn't seem related to this PR. Is this intentional?

Comment (HaoranYi) [reply runtime/src/blockhash_queue.rs:133]: This error shows up in my local clippy run. But it looks like on CI, clippy is happy with that. 
```
> cargo clippy -Zunstable-options -p solana-runtime -- --deny=warnings --deny=clippy::integer_arithmetic --no-deps
    Blocking waiting for file lock on build directory
   Compiling solana-runtime v1.10.1 (D:\mydoc\git\hy_solana\runtime)
error: this expression borrows a value the compiler would automatically borrow
   --> runtime\src\blockhash_queue.rs:133:9
    |
133 |         (&self.ages).iter().map(|(k, v)| {
    |         ^^^^^^^^^^^^ help: change this to: `self.ages`
    |
    = note: `-D clippy::needless-borrow` implied by `-D warnings`
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#needless_borrow

error: could not compile `solana-runtime` due to previous error
```
wonder if you have any idea for this?

Comment (brooksprumo) [reply runtime/src/blockhash_queue.rs:133]: I would just remove this change from the PR.

You can keep the change locally to appease clippy, or open another PR for just that change too. This way it'll keep this PR as small as possible.

File: download-utils/src/lib.rs [hunk 7]
```diff
@@ -268,6 +268,11 @@ pub fn download_snapshot_archive<'a, 'b>(
         maximum_incremental_snapshot_archives_to_retain,
     );
 
+    let snapshot_archives_remote_dir = snapshot_archives_dir
+        .to_path_buf()
+        .join(snapshot_utils::SNAPSHOT_ARCHIVE_DOWNLOAD_DIR);
```
Comment (brooksprumo) [inline download-utils/src/lib.rs:273]: Since this logic is going to be repeated in a few places, how about making it into its own function in snapshot_utils? Something like:

```rust
pub fn build_snapshot_archives_remote_dir(snapshot_archives_dir: impl Into<PathBuf>) -> PathBuf {
    snapshot_archives_dir
        .into()
        .join(SNAPSHOT_ARCHIVE_DOWNLOAD_DIR)
}
```

And then it called be called with a bunch of different inputs:

```rust
        let _ = build_snapshot_archives_remote_dir("string literal");

        let x = "str in var";
        let _ = build_snapshot_archives_remote_dir(x);

        let x = &Path::new("path borrowed");
        let _ = build_snapshot_archives_remote_dir(x);

        let x = Path::new("path moved");
        let _ = build_snapshot_archives_remote_dir(x);

        let x = &PathBuf::from_str("pathbuf borrowed").unwrap();
        let _ = build_snapshot_archives_remote_dir(x);

        let x = PathBuf::from_str("pathbuf moved").unwrap();
        let _ = build_snapshot_archives_remote_dir(x);
```

Comment (HaoranYi) [reply download-utils/src/lib.rs:273]: Good idea.

**Comment by HaoranYi** [comment]
@sakridge Manual tests with and without this change. 
logs w/o this change:
```
[2022-03-09T20:49:56.614354505Z INFO  solana_runtime::snapshot_utils] rebuild bank from snapshots took 81.8s
[2022-03-09T20:49:56.614406930Z INFO  solana_runtime::bank] cleaning..
...
[2022-03-09T20:54:32.785820428Z INFO  solana_runtime::bank] verify_bank_hash..
```
logs with this change:
```
[2022-03-09T21:23:21.566019957Z INFO  solana_runtime::snapshot_utils] rebuild bank from snapshots took 75.9s
[2022-03-09T21:23:21.567471112Z INFO  solana_runtime::bank] verify_bank_hash..
```
It shows about 5 minutes saving on restart time.

**Review by sakridge — DISMISSED** [review]
lgtm, thanks

File: runtime/src/snapshot_utils.rs [hunk 8]
```diff
@@ -1136,54 +1143,58 @@ pub(crate) fn parse_incremental_snapshot_archive_filename(
...
```
Comment (brooksprumo) [inline runtime/src/snapshot_utils.rs:1147]: I still feel like this fn can be simplified, and made more safe. Some things:

- At line 1161, `p.ends_with(SNAPSHOT_ARCHIVE_DOWNLOAD_DIR)`, I don't like the "ends_with" check. Since that means there could be other directories that incorrectly get read here, i.e. `backup_remote`.

- We get the snapshot archives dir from the snapshot config, and we have a fn `build_snapshot_archives_remote_dir()` to get the remote directory. I think `build_snapshot_archives_remote_dir()` should be used, and not reimplement that logic in this function.

- I think overall it'd be simpler to get all the snapshots in the snapshot archives dir, and then get all the snapshots in the remote dir. Walking the directory recursively seems unnecessary, and potentially an issue if somehow there are `remote/remote/remote/remote/../remote/` directories with the same or duplicated snapshots. 

- I'd like to see this fn return the snapshot archive infos instead of using a `&mut` parameter.

**Review by brooksprumo — COMMENTED** [review]
Looking good, and very close! A few thoughts and nits, and then I think it'll be good-to-go!

File: runtime/src/snapshot_utils.rs (hunk 8 above)
Comment (HaoranYi) [reply runtime/src/snapshot_utils.rs]: yup. Fixed.

File: runtime/src/snapshot_utils.rs [hunk 9]
```diff
@@ -1136,54 +1143,57 @@ pub(crate) fn parse_incremental_snapshot_archive_filename(
...
+    F: Fn(PathBuf) -> Result<T>,
+{
+    let walk_dir = |dir| -> Vec<T> {
+        let entry_iter = fs::read_dir(dir);
+        match entry_iter {
+            Err(err) => {
+                info!(
+                    "Unable to read snapshot archives directory: err: {}, path: {}",
+                    err,
+                    snapshot_archives_dir.display()
```
Comment (brooksprumo) [inline runtime/src/snapshot_utils.rs:1158]: Should this be using the `dir` parameter instead?

```suggestion
                    dir.display()
```

**Review by brooksprumo — COMMENTED** [review]
Thanks, I like the new impl!

**Review by brooksprumo — APPROVED** [review]
Lgtm! Thanks for sticking with all my requests 😊

**Review by jeffwashington — APPROVED** [review]
lgtm
